src/quest/serializer.py

```diff
@@ -0,0 +1,65 @@
+from typing import Any, Dict, Tuple
+import importlib
+
+
+class MasterSerializer:
+    def serialize(self, obj: Any) -> Any:
+        # Check if it is a known type - directly serializable to JSON
+        if isinstance(obj, (int, float, str, bool, type(None))):
+            return obj
+        elif isinstance(obj, dict):
+            return {self.serialize(k): self.serialize(v) for k, v in obj.items()}
+        elif isinstance(obj, list):
+            return [self.serialize(v) for v in obj]
+        elif isinstance(obj, tuple):
+            return tuple(self.serialize(v) for v in obj)
+        # Unknown type (Object) - serialize using type and constructor arguments
+        else:
+            args, kwargs = self._get_constructor_args(obj)
+            return {
+                '_ms_type': obj.__module__ + '.' + obj.__class__.__name__,
+                'args': self.serialize(args),
+                'kwargs': self.serialize(kwargs)
+            }
```

Comment:
My original vision for this was that the user (developer) would register known type serializers (which could be class constructors, but could also be other functions):
```py
class MasterSerializer:
    def __init__(self, type_serializers: dict):
        self._type_serializers = type_serializers
        
    ...
    
    def serialize(self, obj):
        ...
        else:
            factory, args, kwargs = self._type_serializers.get(type(obj))
            return {'_ms_type': get_factory_type(factory), 'args': args, 'kwargs': kwargs }
```
a serializer would look like this:
```py
async def load_message(channel_id, message_id):
    channel = await client.fetch_channel(channel_id)
    message = await channel.fetch_message(message_id)
    return message

def serialize_message(message: Message):
    return load_message, (message.channel.id, message.id), {}

...
master_serializer = MasterSerializer({discord.Message: serialize_message})
```

Reply:
You'll have to discern whether a given serializer is async or not (look at `handle_step` for an example). This means that `serialize` and `deserialize` will both have to be async.

Reply:
When serializing and deserializing: 
If the factory is a *function*, the name will look different than if it is a *class*. We should support both. 
But if you store the right thing for `_ms_type`, then I think the loading logic should be the same in either case.